frontend/main.py

```diff
@@ -53,11 +62,92 @@ def aws_cognito_redirect():
+    if 'action' not in req_data and 'corrected_hadith' not in req_data:
+        return jsonify(create_response_message("Please provide valid action param 'delete' or 'approve' and 'corrected_hadith' param"))
+
+    action = req_data['action']
+    corrected_hadith = req_data['corrected_hadith']
+    username = request.cookies.get('username')
+
+    if action == "delete":
+        return archive_correction(correction_id, username, None, False)
+
+    elif action == "approve":
+        dynamodb = boto3.resource('dynamodb',
+            endpoint_url=app.config['DYNAMODB_ENDPOINT_URL'],
+            region_name=app.config['REGION'])
+        table = dynamodb.Table(app.config['DYNAMODB_TABLE'])
+        try:
+            response = table.get_item(Key={'id': str(correction_id)})
+            if 'Item' not in response:
+                return jsonify(create_response_message("correction with id " + str(correction_id) + " not found"))
+ 
+            rows_affected = save_correction_to_hadith_table(response['Item']['urn'], corrected_hadith)
+
+            if rows_affected == 1:
+                archive_correction(correction_id, username, corrected_hadith, True)
+                return jsonify(create_response_message("Successfully updated hadith text"))
+            else:
+                return jsonify(create_response_message("Failed to update hadith text"))
+
+        except ClientError as e:
+            print(e)
+            return jsonify(create_response_message(e.response['Error']['Message']))
+        except pymysql.Error as error:
+            print(error)
+            return jsonify(create_response_message(str(error)))
+        except Exception as exception:
+            return jsonify(create_response_message("Error - " + str(exception)))
+
+    else:
+        return jsonify(create_response_message("Please provide valid action param 'delete' or 'approve'"))
 
 @app.route('/sign_in')
 def sign_in():
     return redirect(aws_auth.get_sign_in_url())
 
+def save_correction_to_hadith_table(urn, corrected_hadith):
+    conn = pymysql.connect(**mysql_properties)
+    cursor = conn.cursor()
+    query = "UPDATE bukhari_english SET hadithText = %(hadith_text)s WHERE englishURN = %(urn)s;"
+    cursor.execute(query, { "hadith_text": corrected_hadith, "urn": urn })
+    rows_affected = cursor.rowcount
+    conn.commit()
+    conn.close()
+    return rows_affected
+
+# Will archive (log) an approved or deleted correction to dynamodb
+def archive_correction(correction_id, username, corrected_hadith=None, approved=False):
+    dynamodb = boto3.resource('dynamodb',
+                endpoint_url=app.config['DYNAMODB_ENDPOINT_URL'],
+                region_name=app.config['REGION'])
+    table = dynamodb.Table(app.config['DYNAMODB_TABLE'])
+    archive_table = dynamodb.Table(app.config['DYNAMODB_TABLE_ARCHIVE'])
+    try:
+        response = table.get_item(Key={'id': str(correction_id)})
+        archive_response = archive_table.put_item(Item={
```

Comment:
please factor out code for adding the archive entry into a separate method, code should read like 

read item
archive item
delete item